src/execution/execute.js

```diff
@@ -222,10 +256,350 @@ function executeOperation(
+  fieldASTs: Array<Field>,
+  fieldName: string,
+  parentType: GraphQLCompositeType
+): GraphQLSerializationPlan {
+  invariant(returnType.serialize, 'Missing serialize method on type');
+
+  const plan: GraphQLSerializationPlan = {
+    kind: 'serialize',
+    fieldASTs,
+    returnType,
+    fieldName,
+    parentType
+  };
+
+  return plan;
+}
+
+/**
+ * Plans the evaluation of a GraphQLList value during value completion.
+ */
+function planMapping(
+  exeContext: ExecutionContext,
+  returnType: GraphQLList,
+  fieldASTs: Array<Field>,
+  fieldName: string,
+  parentType: GraphQLCompositeType
+): GraphQLMappingPlan {
+  const innerType = returnType.ofType;
+
+  const elementPlan =
+    planCompleteValue(
+      exeContext,
+      innerType,
+      fieldASTs,
+      fieldName,
+      parentType
+    );
+
+  const plan: GraphQLMappingPlan = {
+    kind: 'map',
+    fieldASTs,
+    returnType,
+    fieldName,
+    parentType,
+    listElement: elementPlan
+  };
+
+  return plan;
+}
+
+/**
+ * Plan the evaluation of a GraphQLObjectType value during value completion.
+ */
+function planSelection(
+  exeContext: ExecutionContext,
+  type: GraphQLObjectType,
+  fieldASTs: Array<Field>,
+  fieldName: string,
+  parentType: GraphQLCompositeType
+): GraphQLSelectionPlan {
+
+  let fields = Object.create(null);
+  const visitedFragmentNames = Object.create(null);
+  for (let i = 0; i < fieldASTs.length; i++) {
+    const selectionSet = fieldASTs[i].selectionSet;
+    if (selectionSet) {
+      fields = collectFields(
+        exeContext,
+        type,
+        selectionSet,
+        fields,
+        visitedFragmentNames
+      );
+    }
   }
-  return executeFields(exeContext, type, rootValue, fields);
+
+  const {fieldPlansByAlias, fieldPlans} = planFields(exeContext, type, fields);
+
+  const plan: GraphQLSelectionPlan = {
+    kind: 'select',
+    fieldName,
+    fieldASTs,
+    returnType: type,
+    parentType,
+    schema: exeContext.schema,
+    fragments: exeContext.fragments,
+    rootValue: exeContext.rootValue,
+    operation: exeContext.operation,
+    variableValues: exeContext.variableValues,
+    fields: fieldPlans,
+    fieldPlansByAlias
+  };
+
+  return plan;
+}
+
+/**
+ * Plans the evaluation of an abstract type
+ * value during value completion.
+ */
+function planCoercion(
+  exeContext: ExecutionContext,
+  returnType: GraphQLAbstractType,
+  fieldASTs: Array<Field>,
+  fieldName: string,
+  parentType: GraphQLCompositeType
+): GraphQLCoercionPlan {
+  const abstractType = ((returnType: any): GraphQLAbstractType);
+  const possibleTypes = abstractType.getPossibleTypes();
+  const typeChoices = Object.create(null);
+  possibleTypes.forEach(possibleType => {
+    invariant(
+      !typeChoices[possibleType.name],
+      'Two types cannot have the same name "${possibleType.name}"' +
+      'as possible types of abstract type ${abstractType.name}'
+    );
+    typeChoices[possibleType.name] = planSelection(
```

Comment:
It might make sense to calculate `typeChoices` lazily, otherwise number of nodes in a plan tree might grow exponentially with each nesting level of `GraphQLInterface`.